File: Java/le2/plp/Exemplo1.java
package le2.plp;

import java.util.ArrayList;
import java.util.List;

import le2.plp.expressions2.Programa;
import le2.plp.expressions2.declaration.DecVariavel;
import le2.plp.expressions2.expression.ExpDeclaracao;
import le2.plp.expressions2.expression.ExpSoma;
import le2.plp.expressions2.expression.Expressao;
import le2.plp.expressions2.expression.Id;
import le2.plp.expressions2.expression.ValorInteiro;

public class Exemplo1 {

    static void executar(String descricao, Programa prog) {
        System.out.println("--------------------------------------------------");
        System.out.println("Expressao : " + descricao);
        try {
            boolean bemTipada = prog.checaTipo();
            System.out.println("Bem tipada: " + bemTipada);
            if (bemTipada) {
                System.out.println("Resultado: " + prog.executar());
            } else {
                System.out.println("Erro de tipo: avaliacao abortada.");
            }
        } catch (Exception e) {
            System.out.println("Excecao capturada: " + e.getMessage());
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        DecVariavel decX = new DecVariavel(new Id("x"), new ValorInteiro(1));
        List<DecVariavel> decs = new ArrayList<DecVariavel>();
        decs.add(decX);
        
        Expressao corpo  = new ExpSoma(new Id("x"), new ValorInteiro(1));
        Expressao let    = new ExpDeclaracao(decs, corpo);
        Programa prog    = new Programa(let);

        executar("let var x = 1 in x + 1", prog);

        System.out.println("--------------------------------------------------");
    }
}

File: Python/le2-python/Exemplo1.py
from Programa import Programa
from declaration.dec_variavel import DecVariavel
from expression.exp_declaracao import ExpDeclaracao
from expression.exp_soma import ExpSoma
from expression.id import Id
from expression.valor_inteiro import ValorInteiro


def executar(descricao: str, prog: Programa) -> None:
    print("--------------------------------------------------")
    print("Expressao : " + descricao)
    try:
        bem_tipada: bool = prog.checa_tipo()
        print("Bem tipada: " + str(bem_tipada).lower())
        if bem_tipada:
            print("Resultado: " + str(prog.executar()))
        else:
            print("Erro de tipo: avaliacao abortada.")
    except Exception as e:
        print("Excecao capturada: " + str(e))
        import traceback

        traceback.print_exc()

def main():
    # DecVariavel("x", 1)
    dec_x = DecVariavel(Id("x"), ValorInteiro(1))
    decs = [dec_x]

    # corpo: x + 1   |   let: let var x = 1 in corpo
    corpo = ExpSoma(Id("x"), ValorInteiro(1))
    let = ExpDeclaracao(decs, corpo)
    prog = Programa(let)

    executar("let var x = 1 in x + 1", prog)

    print("--------------------------------------------------")



if __name__ == "__main__":
    main()